### PR TITLE
CWaterCreatures

### PR DESCRIPTION
**Pull request by majesticCoding**
commented and not tested because it has conflicts with CDummy and CPopulation (looks like something isn't finished about these classes)

### CODE REVIEW

File: src/render/WaterCreatures.cpp [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		return;
+
+	CVector playerPos = FindPlayerPed()->GetPosition();
+	for (int i = 0; i < NUM_WATER_CREATURES; i++) {
+		switch (aWaterCreatures[i].m_state) {
+		case WATER_CREATURE_ALLOCATED:
+		case WATER_CREATURE_ACTIVE: {
+			aWaterCreatures[i].m_pObj->m_nEndOfLifeTime = CTimer::GetTimeInMilliseconds() + 40000;
+			if (aWaterCreatures[i].m_pObj->GetIsOnScreen()) {
+				if ((playerPos - aWaterCreatures[i].m_pObj->GetPosition()).Magnitude() < 5625.0f) {
```
Comment (Sergeanur) [inline src/render/WaterCreatures.cpp:195]: SQR(75.0f)

Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+void CWaterCreatures::UpdateAll() {
+	if (nNumActiveSeaLifeForms == 0)
+		return;
+
+	CVector playerPos = FindPlayerPed()->GetPosition();
+	for (int i = 0; i < NUM_WATER_CREATURES; i++) {
+		switch (aWaterCreatures[i].m_state) {
+		case WATER_CREATURE_ALLOCATED:
+		case WATER_CREATURE_ACTIVE: {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:192]: these are obviously not equal cases

File: src/render/WaterCreatures.cpp [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+			aWaterCreatures[i].m_pObj->m_nEndOfLifeTime = CTimer::GetTimeInMilliseconds() + 40000;
+			if (aWaterCreatures[i].m_pObj->GetIsOnScreen()) {
+				if ((playerPos - aWaterCreatures[i].m_pObj->GetPosition()).Magnitude() < SQR(75.0f)) {
+					//if (aWaterCreatures[i].m_alpha < 255)
+						//aWaterCreatures[i].m_alpha = Min(aWaterCreatures[i].m_alpha + 4, 255);
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:197]: you're missing SetRwObjectAlpha in here?

File: src/render/WaterCreatures.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+						if (aWaterCreatures[i].m_pObj->m_nEndOfLifeTime - 40000 <= CTimer::GetTimeInMilliseconds())
+							aWaterCreatures[i].m_state = WATER_CREATURE_ACTIVE;
+					}
+					else {
+						aWaterCreatures[i].m_state = WATER_CREATURE_UPDATE;
+					}
+
+				}
+				else {
+					aWaterCreatures[i].m_state = WATER_CREATURE_TO_REMOVE;
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:215]: is it intentional that you didn't add SetRwObjectAlpha in here?

File: src/render/WaterCreatures.cpp [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+		}
+		case WATER_CREATURE_TO_REMOVE: {
+			if (aWaterCreatures[i].m_pObj)
+				CWorld::Remove(aWaterCreatures[i].m_pObj);
+			FreeFishStructSlot(&aWaterCreatures[i]);
+			nNumActiveSeaLifeForms--;
+			aWaterCreatures[i].m_state = WATER_CREATURE_DISABLED;
+			break;
+		}
+		case WATER_CREATURE_DISABLED:
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:248]: this is `default`,  `WATER_CREATURE_DISABLED`

File: src/render/WaterCreatures.cpp (hunk 3 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp:197]: commented now, because it gives an error

File: src/render/WaterCreatures.cpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+					//if (aWaterCreatures[i].m_alpha < 255)
+						//aWaterCreatures[i].m_alpha = Min(aWaterCreatures[i].m_alpha + 4, 255);
+					CVector newRight = aWaterCreatures[i].m_pObj->GetRight();
+					newRight.Normalise();
+					aWaterCreatures[i].m_pObj->m_vecMoveSpeed = newRight * aWaterCreatures[i].m_fRightMult;
+					aWaterCreatures[i].m_pObj->m_vecTurnSpeed = CVector(0.0f, 0.0f, aWaterCreatures[i].m_fZTurnSpeed);
+					aWaterCreatures[i].m_pObj->bIsStatic = false;
+					float fDepth = 0.0;
+					CWaterLevel::GetWaterDepth(aWaterCreatures[i].m_pObj->GetPosition(), &fDepth, nil, nil);
+					if (fDepth < 0.0f) {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:205]: operand and right-hand side is wrong

File: src/render/WaterCreatures.cpp [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+void CWaterCreatures::CreateOne(CVector pos, int16 modelID) {
+	if (!IsSpaceForMoreWaterCreatures())
+		return;
+
+	float fDepth, fLevelNoWaves;
+	if (!TheCamera.IsSphereVisible(pos, 3.0f) 
+		&& CWaterLevel::GetWaterDepth(pos, &fDepth, &fLevelNoWaves, nil) && fDepth > 4.5f) {
+		
+		if (modelID == -1 || modelID > 64)
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:162]: you're missing the modelID < 0 condition. original code should be looking like that;
```
	int modelToUse = -1;
	if ( modelID != -1 && modelID >= 0 && modelID <= 64 )
		modelToUse = modelID;

	if ( modelToUse == -1 )
		modelToUse = CGeneral::GetRandomNumberInRange(0, 64);
```

but if you want to keep your 'style', then you can do it
```
if (modelID == -1 || modelID < 0 || modelID > 64) 
```

File: src/render/WaterCreatures.cpp [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+	float fDepth, fLevelNoWaves;
+	if (!TheCamera.IsSphereVisible(pos, 3.0f) 
+		&& CWaterLevel::GetWaterDepth(pos, &fDepth, &fLevelNoWaves, nil) && fDepth > 4.5f) {
+		
+		if (modelID == -1 || modelID > 64) 
+			modelID = CGeneral::GetRandomNumberInRange(0, 64);
+
+		WaterCreatureProperties *creature = &aProperties[modelID];
+		pos.z = fLevelNoWaves - creature->fLevel;
+		float fRightMult = CGeneral::GetRandomNumberInRange(0.0f, 0.01f);
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:167]: it should be multiplied with `creature->fRightMult`

Comment (majesticCoding) [reply src/render/WaterCreatures.cpp:167]: hmm, if I do it, fishes just appear higher than the water level and become static

Comment (erorcun) [reply src/render/WaterCreatures.cpp:167]: yeah sorry, it should be `CGeneral::GetRandomNumberInRange(0.0f, creature->fRightMult) + 0.01f`

Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp (hunk 7 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp (hunk 6 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp (hunk 5 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp (hunk 2 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: fixed

File: src/render/WaterCreatures.cpp (hunk 4 above)
Comment (majesticCoding) [reply src/render/WaterCreatures.cpp]: added

File: src/render/WaterCreatures.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+CWaterCreature *CWaterCreatures::GetFishStructSlot() {
+	for (int i = 0; i < NUM_WATER_CREATURES; i++)
+		if (aWaterCreatures[i].m_state == WATER_CREATURE_DISABLED)
+			return &aWaterCreatures[i];
+	
+	return nil;
+}
+
+CObject *CWaterCreatures::CreateSeaLifeForm(CVector pos, int16 *modelID, float zRotAngle) {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:114]: please use the original function signature

File: src/render/WaterCreatures.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+float CWaterCreatures::CalculateFishHeading(CVector const& pos1, CVector const& pos2) {
+	CVector delta = pos1 - pos2;
+	delta.Normalise();
+
+	return Atan2(-delta.x, delta.y);
+}
+
+void CWaterCreatures::CreateOne(CVector pos, int16 modelID) {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:154]: same

File: src/render/WaterCreatures.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+CWaterCreature::CWaterCreature() {
+	Free();
+}
+
+CWaterCreature::~CWaterCreature() {
+	//looks like unused
+}
+
+void CWaterCreature::Initialise(CObject *pObj, float fRightMult, float fZTurnSpeed, float fWaterDepth, uint32 alpha, int eFishSlotState) {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:89]: same

File: src/render/WaterCreatures.cpp [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+	this->m_pObj = pObj;
+	this->m_fRightMult = fRightMult;
+	this->m_fZTurnSpeed = 0.0f;
+	this->m_fWaterDepth = fWaterDepth;
+	this->m_alpha = alpha;
+	this->m_state = eFishSlotState;
+}
+
+void CWaterCreature::Allocate(CObject *pObj, float fRightMult, float fZTurnSpeed, float fWaterDepth, uint32 alpha, int eFishSlotState) {
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:98]: same

File: src/render/WaterCreatures.cpp [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+CWaterCreature::~CWaterCreature() {
+	//looks like unused
+}
+
+void CWaterCreature::Initialise(CObject *pObj, float fRightMult, float fZTurnSpeed, float fWaterDepth, uint32 alpha, int eFishSlotState) {
+	this->m_pObj = pObj;
+	this->m_fRightMult = fRightMult;
+	this->m_fZTurnSpeed = 0.0f;
```
Comment (erorcun) [inline src/render/WaterCreatures.cpp:92]: looks like you forgot debugging thing here

**Review by erorcun — APPROVED** [review]

**Review by Nick007J — APPROVED** [review]

**Pull request merged**